### Game data exploration

ARRM database files were copied from ARRM install folder and converted from mdb to csv using mdb-tools

Sources:
- ARRM database files
- SteamDB data
- what else?

Preferred features for a database:
- name/title
- filename (important for arcade titles)
- description
- platform
- release date/year
- genre(s)
- developer
- publisher
- number of players
- cooperative
- rating

In [38]:
# !pip install pandas

In [39]:
import pandas as pd

## DAT Databases

In [40]:
dat_daphne = pd.read_csv('csv/dat_database_daphne.csv')
dat_dos = pd.read_csv('csv/dat_database_dos.csv')
dat_mame = pd.read_csv('csv/dat_database_mame.csv')
dat_scumm_vm = pd.read_csv('csv/dat_database_scummvm.csv')

/tmp/ipykernel_11215/1416037915.py:3: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  dat_mame = pd.read_csv('csv/dat_database_mame.csv')


### DOS

In [41]:
# display dat_dos info
dat_dos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7482 entries, 0 to 7481
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                7482 non-null   int64 
 1   gametitle_mame    7482 non-null   object
 2   developer_mame    7479 non-null   object
 3   year_mame         7482 non-null   object
 4   genre_mame        7481 non-null   object
 5   description_mame  7470 non-null   object
 6   publisher_mame    7451 non-null   object
dtypes: int64(1), object(6)
memory usage: 409.3+ KB


In [42]:
dat_dos.isnull().sum()

ID                   0
gametitle_mame       0
developer_mame       3
year_mame            0
genre_mame           1
description_mame    12
publisher_mame      31
dtype: int64

In [43]:
dat_dos.head()

ID   gametitle_mame         developer_mame        year_mame  \
0  61191     Stunt Island      The Assembly Line  19920101T000000   
1  61192  Tommy's Yahtzee           Tommy's Toys  19860101T000000   
2  61193  Fantasy Empires        Silicon Knights  19931001T000000   
3  61194             Pune  Hardware Not Included  19950101T000000   
4  61195    Mortal Kombat    Probe Software Ltd.  19940525T000000   

                            genre_mame  \
0  Flight Simulator,Vehicle Simulation   
1                   Board / Party Game   
2                             Strategy   
3                               Action   
4                      Action,Fighting   

                                    description_mame  \
0  Stunt Island was marketed as The Stunt Flying ...   
1  Tommy's Yahtzee is a shareware implementation ...   
2  Build and control an Empire! In Fantasy Empire...   
3  Pune is a light cycle/snake-type game for mult...   
4  Five Hundred years ago, an ancient and well re...   

                        publisher_mame  
0  Walt Disney Computer Software, Inc.  
1                             Freeware  
2          Strategic Simulations, Inc.  
3                Hardware Not Included  
4          Acclaim Entertainment, Inc.

DOS analysis: A very nice set and i love DOS games. Lets use it.

### Daphne

In [44]:
dat_daphne.head()

ID    gametitle_mame  filename_mame  cloneof_mame  manufacturer_mame  \
0  599  .38 Ambush Alley  38ambushalley           NaN                NaN   
1  600         Space Ace            ace           NaN                NaN   
2  601         Space Ace          ace91           NaN                NaN   
3  602         Space Ace     ace91_euro           NaN                NaN   
4  603         Space Ace        aceeuro           NaN                NaN   

  year_mame  genre_mame  
0         -         NaN  
1         -         NaN  
2         -         NaN  
3         -         NaN  
4         -         NaN

In [45]:
dat_daphne.isnull().sum()

ID                     0
gametitle_mame         0
filename_mame          0
cloneof_mame         208
manufacturer_mame    208
year_mame              0
genre_mame           208
dtype: int64

Analysis: Daphne dataframe is only 34 rows, nulls for manufacturer and genre info. Lets skip this database.

### MAME

In [46]:
dat_mame.head()

ID                systemname_mame  \
0  355065  advmame 0.94-RetroPie-260.dat   
1  355066  advmame 0.94-RetroPie-260.dat   
2  355067  advmame 0.94-RetroPie-260.dat   
3  355068  advmame 0.94-RetroPie-260.dat   
4  355069  advmame 0.94-RetroPie-260.dat   

                            gametitle_mame filename_mame cloneof_mame  \
0  PuckMan (Japan set 1, Probably Bootleg)       puckman          NaN   
1                    PuckMan (Japan set 2)      puckmana      puckman   
2  PuckMan (Japan set 1 with speedup hack)      puckmanf      puckman   
3                         Puckman (Falcom)      puckmanh      puckman   
4                         Pac-Man (Midway)        pacman      puckman   

          manufacturer_mame   year_mame genre_mame  
0                     Namco  1980/01/01        NaN  
1                     Namco  1980/01/01        NaN  
2                     Namco  1980/01/01        NaN  
3                      hack  1980/01/01        NaN  
4  [Namco] (Midway license)  1980/01/01        NaN

In [47]:
dat_mame.isnull().sum()

ID                        0
systemname_mame           0
gametitle_mame            0
filename_mame             0
cloneof_mame         101633
manufacturer_mame     11907
year_mame             95935
genre_mame           193224
dtype: int64

Analysis: The MAME list is huge (222251 rows), genre missing for 87%, year for 42% of the titles. No descriptions. However the existence of mame filenames could be useful. <br>
Columns to keep: gametitle_mame, filename_mame manufacturer_mame and year_mame.

### ScummVM

In [48]:
dat_scumm_vm.head()

ID                             gametitle_mame filename_mame  \
0  44251  The 11th Hour The Sequel to The 7th Guest           11h   
1  44252                         Five Lethal Demons           5ld   
2  44253                       Five Magical Amulets           5ma   
3  44254                 Amazon - Guardians Of Eden        access   
4  44255   Putt-Putt And Fatty Bear's Activity Pack      activity   

   cloneof_mame        manufacturer_mame year_mame genre_mame  
0           NaN                Trilobyte         -        NaN  
1           NaN               Off Studio         -        NaN  
2           NaN               Off Studio         -        NaN  
3           NaN                      NaN         -        NaN  
4           NaN  Humongous Entertainment         -        NaN

In [49]:
dat_scumm_vm.isnull().sum()

ID                     0
gametitle_mame         0
filename_mame          0
cloneof_mame         483
manufacturer_mame    149
year_mame              0
genre_mame           318
dtype: int64

SCUMM analysis: Bit of a niche dataset but still useful. Manufacturer and genre missing for 35% titles but after inspection the 35% represents more niche games or some non-SCUMM games. Let's come back to this later.

## GameTDB

In [50]:
gamestdb = pd.read_csv('csv/games_on_gametdb.csv')

In [51]:
gamestdb.head()

N° gametdb_id gametdb_type gametdb_region        gametdb_languages  \
0  1108862       A22J          3DS         NTSC-J                       JP   
1  1108863       A2AA          3DS            NaN  EN,FR,DE,ES,IT,JP,KR,TW   
2  1108864       A2AE          3DS         NTSC-U                       EN   
3  1108865       A2AJ          3DS         NTSC-J                       JP   
4  1108866       A2AK          3DS         NTSC-K                       KR   

                     gametdb_title_de                    gametdb_title_en  \
0  Bokujou Monogatari Futago no Mura+  Bokujou Monogatari Futago no Mura+   
1          Pokémon Ultrasonne: Update                   Pokémon Ultra Sun   
2                   Pokémon Ultra Sun                   Pokémon Ultra Sun   
3                   Pokémon Ultra Sun                   Pokémon Ultra Sun   
4                   Pokémon Ultra Sun                   Pokémon Ultra Sun   

                           gametdb_title_es  \
0                                       NaN   
1  Pokémon Ultra Sun Actualización Ver. 1.2   
2                                       NaN   
3                                       NaN   
4                                       NaN   

                     gametdb_title_fr                    gametdb_title_it  \
0  Bokujou Monogatari Futago no Mura+  Bokujou Monogatari Futago no Mura+   
1   Pokémon Ultra-Soleil: Mise à jour    Pokémon Ultrasole: Aggiornamenti   
2                   Pokémon Ultra Sun                   Pokémon Ultra Sun   
3                   Pokémon Ultra Sun                   Pokémon Ultra Sun   
4                   Pokémon Ultra Sun                   Pokémon Ultra Sun   

   ... gametdb_players                      gametdb_genre  \
0  ...             1.0                                NaN   
1  ...             1.0            role-playing,action rpg   
2  ...             1.0  adventure,role-playing,action rpg   
3  ...             1.0  adventure,role-playing,action rpg   
4  ...             1.0            role-playing,action rpg   

                                         gametdb_rom  \
0  Bokujou Monogatari Futago no Mura+ (Japan) (JA...   
1  Pokémon Ultra Sun (EN,FR,DE,ES,IT,JA,KO,ZHTW).3ds   
2                   Pokémon Ultra Sun (USA) (EN).3ds   
3                 Pokémon Ultra Sun (Japan) (JA).3ds   
4                 Pokémon Ultra Sun (Korea) (KO).3ds   

                              gametdb_rom_sans_ext gametdb_platform  \
0  Bokujou Monogatari Futago no Mura+ (Japan) (JA)              3DS   
1    Pokémon Ultra Sun (EN,FR,DE,ES,IT,JA,KO,ZHTW)              3DS   
2                     Pokémon Ultra Sun (USA) (EN)              3DS   
3                   Pokémon Ultra Sun (Japan) (JA)              3DS   
4                   Pokémon Ultra Sun (Korea) (KO)              3DS   

                     gametdb_title_cn                    gametdb_title_tw  \
0  Bokujou Monogatari Futago no Mura+  Bokujou Monogatari Futago no Mura+   
1                   Pokémon Ultra Sun            更新資料 Ver. 1.2 精靈寶可夢 究極之日   
2                   Pokémon Ultra Sun                   Pokémon Ultra Sun   
3                   Pokémon Ultra Sun                   Pokémon Ultra Sun   
4                   Pokémon Ultra Sun                   Pokémon Ultra Sun   

                                 gametdb_synopsis_cn  \
0                                                NaN   
1  An update for Pokémon™ Ultra Sun is now availa...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                 gametdb_synopsis_tw  \
0                                                NaN   
1  Nintendo 3DS遊戲軟體『精靈寶可夢 究極之日』的更新資料。\n持有『精靈寶可夢 究...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                  gametdb_rom_cleaned  
0  Bokujou Monogatari 

In [52]:
gamestdb.columns

Index(['N°', 'gametdb_id', 'gametdb_type', 'gametdb_region',
       'gametdb_languages', 'gametdb_title_de', 'gametdb_title_en',
       'gametdb_title_es', 'gametdb_title_fr', 'gametdb_title_it',
       'gametdb_title_ja', 'gametdb_title_ko', 'gametdb_title_nl',
       'gametdb_title_pt', 'gametdb_title_ru', 'gametdb_synopsis_de',
       'gametdb_synopsis_en', 'gametdb_synopsis_es', 'gametdb_synopsis_fr',
       'gametdb_synopsis_it', 'gametdb_synopsis_ja', 'gametdb_synopsis_ko',
       'gametdb_synopsis_nl', 'gametdb_synopsis_pt', 'gametdb_synopsis_ru',
       'gametdb_developer', 'gametdb_publisher', 'gametdb_date',
       'gametdb_players', 'gametdb_genre', 'gametdb_rom',
       'gametdb_rom_sans_ext', 'gametdb_platform', 'gametdb_title_cn',
       'gametdb_title_tw', 'gametdb_synopsis_cn', 'gametdb_synopsis_tw',
       'gametdb_rom_cleaned'],
      dtype='object')

In [53]:
gamestdb.isnull().sum()

N°                          0
gametdb_id                  0
gametdb_type                0
gametdb_region           6400
gametdb_languages        2466
gametdb_title_de          912
gametdb_title_en          918
gametdb_title_es        44811
gametdb_title_fr          848
gametdb_title_it          913
gametdb_title_ja          250
gametdb_title_ko          786
gametdb_title_nl          917
gametdb_title_pt          918
gametdb_title_ru          918
gametdb_synopsis_de     34824
gametdb_synopsis_en     34997
gametdb_synopsis_es     45089
gametdb_synopsis_fr     34696
gametdb_synopsis_it     34935
gametdb_synopsis_ja     33357
gametdb_synopsis_ko     34795
gametdb_synopsis_nl     34922
gametdb_synopsis_pt     34988
gametdb_synopsis_ru     34994
gametdb_developer       30586
gametdb_publisher       16140
gametdb_date                0
gametdb_players          8457
gametdb_genre           21728
gametdb_rom               918
gametdb_rom_sans_ext        6
gametdb_platform            0
gametdb_ti

In [54]:
# show distinct gametdb_type values
gamestdb['gametdb_type'].unique()

array(['3DS', 'DS', 'PS3', 'Switch', 'Wii', 'WiiU'], dtype=object)

Gametdb analysis: nice dataset but only 6 platforms represented. We're still going to use it.

## Launchbox

In [55]:
launchbox = pd.read_csv('csv/launchbox.csv')

/tmp/ipykernel_11215/1639964382.py:1: DtypeWarning: Columns (17,18,24,25) have mixed types. Specify dtype option on import or set low_memory=False.
  launchbox = pd.read_csv('csv/launchbox.csv')


In [56]:
launchbox.head()

N°  DatabaseID                                     Name  ReleaseYear  \
0  6658890       17687               20th Century Video Almanac       1993.0   
1  6658891      219683                                 3D Atlas       1994.0   
2  6658892       25182                           3DO Action Pak       1995.0   
3  6658893      417702  3DO de Shiru Miru Asobu Nakajima Miyuki          NaN   
4  6658894      130908                    3DO Demo Disc Program          NaN   

                                            Overview  MaxPlayers ReleaseType  \
0  In The Best of Our Century, we've taken multim...         1.0    Released   
1  The World Isn't Flat. Why Should Your Atlas Be...         1.0    Released   
2  This is a four-game compilation pack that cont...         1.0    Released   
3                                                NaN         1.0  Unreleased   
4  A white binder with blue silk-screened art. Th...         NaN         NaN   

   Cooperative                                     VideoURL  CommunityRating  \
0            0  https://www.youtube.com/watch?v=dfFT6jqjFq0         3.301418   
1            0  https://www.youtube.com/watch?v=RZ-pph55Ui4         3.500000   
2            0                                          NaN         3.647436   
3            0                                          NaN         3.750000   
4            0                                          NaN         3.330357   

   ... AlternateName_China AlternateName_Europe  AlternateName_France  \
0  ...                 NaN                  NaN                   NaN   
1  ...                 NaN                  NaN                   NaN   
2  ...                 NaN                  NaN                   NaN   
3  ...                 NaN                  NaN                   NaN   
4  ...                 NaN                  NaN                   NaN   

  AlternateName_Germany  AlternateName_Japan AlternateName_Korea  \
0                   NaN                  NaN                 NaN   
1                   NaN                  NaN                 NaN   
2                   NaN                  NaN                 NaN   
3                   NaN  3DOで知る・みる・あそぶ 中島みゆき                 NaN   
4                   NaN                  NaN                 NaN   

  AlternateName_NorthAmerica AlternateName_Spain AlternateName_UnitedStates  \
0                        NaN                 NaN                        NaN   
1                        NaN                 NaN                        NaN   
2                        NaN                 NaN                        NaN   
3                        NaN                 NaN                        NaN   
4                        NaN                 NaN                        NaN   

                         AlternateName_World  
0                                        NaN  
1                                        NaN  
2                                        NaN  
3  3DO de Shiru Miru Asobu - Nakajima Miyuki  
4                                        NaN  

[5 rows x 29 columns]

Launchbox analysis: huge list, some missing values. Prime candidate for the main dataset that other datasets will be added to.

## Mobygames

In [57]:
mobygames = pd.read_csv('csv/mobygames.csv')

In [58]:
mobygames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 316257 entries, 0 to 316256
Data columns (total 7 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   numauto        316257 non-null  int64  
 1   platform_id    316257 non-null  int64  
 2   platform_name  316257 non-null  object 
 3   game_id        316257 non-null  int64  
 4   game_title     316257 non-null  object 
 5   release_date   0 non-null       float64
 6   mobygames_url  316257 non-null  object 
dtypes: float64(1), int64(3), object(3)
memory usage: 16.9+ MB


In [59]:
mobygames.isnull().sum()

numauto               0
platform_id           0
platform_name         0
game_id               0
game_title            0
release_date     316257
mobygames_url         0
dtype: int64

In [60]:
mobygames.head()

numauto  platform_id platform_name  game_id        game_title  \
0  2781406           35           3do       96        Quarantine   
1  2781407           35           3do      111          Virtuoso   
2  2781408           35           3do      165  Princess Maker 2   
3  2781409           35           3do      179   Star Control II   
4  2781410           35           3do      210     Cannon Fodder   

   release_date                                      mobygames_url  
0           NaN  t1H5j95PeMaZo+Cw1R9vKw==hY9Nu9IGxGC1KpiwIcK0pX...  
1           NaN  pskb3XM01ASHZRWZmUS00g==9/zl7gK72ixL8RSmDQYsDq...  
2           NaN  IaammuI4ElexSjqcTb6zAg==LU77bhsLWm+TMdKjgrt5Wv...  
3           NaN  5NOpZrNp7xOTX79DuN4weg==uPkHQmlKhUSynxAzrMmt4b...  
4           NaN  ICShm7ilii6jmpRScaapPw==QXCpV2CiXimgLBnmVrOM6c...

Mobygames analysis: Only platform names and game titles here and release dates are empty. Huge list 316257 entries. A good candidate to attach other info to.

## Recalbox

In [61]:
recalbox = pd.read_csv('csv/recalbox_gamelist.csv')

In [62]:
recalbox.head()

numauto_rom                               nomjeu_rom  \
0        33308                 4-in-1 Fun Pak Volume II   
1        33310                        The Addams Family   
2        33312  Adventure Island II: Aliens in Paradise   
3        33314             The Adventures of Star Saver   
4        33315                                 Aerostar   

                                         fichier_rom  \
0       ./4-in-1 Fun Pak Volume II (USA, Europe).zip   
1                     ./Addams Family, The (USA).zip   
2  ./Adventure Island II - Aliens in Paradise (US...   
3  ./Adventures of Star Saver, The (USA, Europe).zip   
4                       ./Aerostar (USA, Europe).zip   

                                     description_rom  \
0  Now you can have four of the most popular boar...   
1  The Addams Family is a side-scrolling platform...   
2  Jeannie Jungle has been captured by aliens! Sh...   
3  On a routine check through the city, police of...   
4  Aerostar is a top-down, vertically-scrolling s...   

                                           image_rom  rating_rom  \
0  ./images/4-in-1 Fun Pak Volume II (USA, Europe...         0.4   
1              ./images/Addams Family, The (USA).png         0.4   
2  ./images/Adventure Island II - Aliens in Parad...         0.4   
3  ./images/Adventures of Star Saver, The (USA, E...         0.4   
4                ./images/Aerostar (USA, Europe).png         1.0   

         annee_rom   developer_rom    publisher_rom  \
0  19920901T000000   Beam Software        Interplay   
1  19920102T000000           Ocean            Ocean   
2  19930202T000000          Hudson           Hudson   
3  19920101T000000          A-Wave            Taito   
4  19910901T000000  Sigma Pro-Tech  Vic Tokai, Inc.   

                    genre_rom  ... arcadesystemname_rom  gametime_rom  \
0  Variados-Jogo de tabuleiro  ...                  NaN            36   
1                        Ação  ...                  NaN             0   
2                        Ação  ...                  NaN             0   
3             Ação-Plataforma  ...                  NaN             0   
4            Shoot'em Up-Ação  ...                  NaN             0   

   boxback_rom temporary_rom bezel_rom ratio_rom  rotation_rom extra1_rom  \
0          NaN           NaN       NaN       NaN           NaN        NaN   
1          NaN           NaN       NaN       NaN           NaN        NaN   
2          NaN           NaN       NaN       NaN           NaN        NaN   
3          NaN           NaN       NaN       NaN           NaN        NaN   
4          NaN           NaN       NaN       NaN           NaN        NaN   

   famille_rom mode_rom  
0          NaN      NaN  
1          NaN      NaN  
2          NaN      NaN  
3          NaN      NaN  
4          NaN      NaN  

[5 rows x 47 columns]

Recalbox: Looks like its (mostly) NES/Famicom titles. It has descriptions. Looks like the xml version has less missing values. Let's try joining these two before merging them to main.

## all_games.csv
(don't know where its from)

In [63]:
all_games = pd.read_csv('csv/all_games.csv')

In [64]:
all_games.head()

name        platform        release_date  \
0  The Legend of Zelda: Ocarina of Time     Nintendo 64   November 23, 1998   
1              Tony Hawk's Pro Skater 2     PlayStation  September 20, 2000   
2                   Grand Theft Auto IV   PlayStation 3      April 29, 2008   
3                           SoulCalibur       Dreamcast   September 8, 1999   
4                   Grand Theft Auto IV        Xbox 360      April 29, 2008   

                                             summary  meta_score user_review  
0  As a young boy, Link is tricked by Ganondorf, ...          99         9.1  
1  As most major publishers' development efforts ...          98         7.4  
2  [Metacritic's 2008 PS3 Game of the Year; Also ...          98         7.7  
3  This is a tale of souls and swords, transcendi...          98         8.4  
4  [Metacritic's 2008 Xbox 360 Game of the Year; ...          98         7.9

In [65]:
all_games.isnull().sum()

name              0
platform          0
release_date      0
summary         114
meta_score        0
user_review       0
dtype: int64

all_games: really nice list. I think it is metacritic top games. barely any missing values. definitely useful.